Add Datadog Incident Management plugin support

bernardjkim opened the pull request:
This PR adds support for self-hosted Access Request Datadog plugin.

I'll create a separate PR for a few todo items that are also required to support the plugin.
Todo:
- [ ] https://github.com/gravitational/teleport.e/pull/4992
- [ ] https://github.com/gravitational/teleport/pull/46514
- [ ] Add support for automatic approvals through Datadog oncall

changelog: Add support for Access Request Datadog plugin.

bernardjkim committed:
Implement datadog plugin

bernardjkim committed:
Add unit tests

bernardjkim committed:
Merge branch 'master' into bernard/datadog-plugin-impl

zmb3 commented:
I woudl prefer that we named this plugin something like `DatadogIncidentManagement` as this is a separate product (similar to PagerDuty), not Datadog's traditional monitoring solution (which there is also a Teleport integration for)

bernardjkim committed:
Add fallback recipient config

bernardjkim committed:
Rename to Datadog Incident Management

zmb3 commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -6050,8 +6050,8 @@ message PluginSpecV1 {
     PluginSCIMSettings scim = 14;
-    // Settings for the Datadog plugin
-    PluginDatadogSettings datadog = 15;
+    // Settings for the Datadog acess plugin
```
```suggestion
    // Settings for the Datadog Incident Management plugin
```

bernardjkim committed:
Update tests

bernardjkim committed:
Datadog Incident Management

bernardjkim committed:
Update tctl resource plugin command

bernardjkim committed:
Merge branch 'master' into bernard/datadog-plugin-impl

bernardjkim committed:
Typos

bernardjkim committed:
Lint

tigrato commented:
Is it possible to split this PR into smaller PRs?

bernardjkim committed:
Exclude api changes for now

bernardjkim committed:
Set channel size

bernardjkim commented:
Friendly ping @hugoShaka @EdwardDowling @marcoandredinis @tigrato

marcoandredinis commented on integrations/access/datadog/cmd/teleport-datadog/main.go:
```diff
@@ -0,0 +1,98 @@
+	}
+
+	app := datadog.NewDatadogApp(conf)
+	go lib.ServeSignals(app, 15*time.Second)
```
What's the 2nd argument? Can we extract the value into a more meaningful variable name?

marcoandredinis commented on integrations/access/datadog/client.go:
```diff
@@ -0,0 +1,236 @@
+	statusEmitTimeout  = 10 * time.Second
+)
+
+const (
+	IncidentWritePermissions = "incident_write"
+)
```
Can we please add a godoc explaining what this permission allows?
A short sentence and link to https://docs.datadoghq.com/account_management/rbac/permissions/#case-and-incident-management should be enough.

marcoandredinis commented on integrations/access/datadog/client.go:
```diff
@@ -0,0 +1,236 @@
+		return trace.Wrap(err)
+	}
+	for _, permission := range result.Data {
+		if permission.Attributes.Name == IncidentWritePermissions && permission.Attributes.Restricted {
+			return trace.AccessDenied("missing incident_write permissions")
+		}
```
Just a nit, but I think we can break the loop as soon as we find the permission we are looking for.

```suggestion
		if permission.Attributes.Name == IncidentWritePermissions {
			if permission.Attributes.Restricted {
				return trace.AccessDenied("missing incident_write permissions")
			}

			return nil
		}
```

marcoandredinis reviewed (commented):
Most things look good
I want to do a real test before giving the final approval

bernardjkim committed:
Address feedback

- Add PluginShutdownTimeout const
- Support api endpoint configuration
- Add additional godocs/comments

hugoShaka commented on integrations/access/datadog/client.go:
```diff
@@ -0,0 +1,236 @@
+		},
+	}
+	var result IncidentsBody
+	_, err := d.client.NewRequest().
```
Historical context: for azure/msteams we did not use the official API because the SDK depends on the entire universe and it took 30 min + to build the plugin (I tried doing it but gave up).

For other implementations this was done by the martians and they likely walked the easiest/quickest pah.

I don't have an opinion on whether we should use or not their go client. I think the most dangerous things with building our own client are the authentication flow and the failure/retry logic. If those are trivial and the go client adds many deps, rolling our own client makes sense.

In this case maybe drop a comment saying why we decided to do it ourselves.

bernardjkim replied:
I'll leave a todo to migrate to the official datadog package if the need arises.

bernardjkim committed:
Comment about datadog client package

marcoandredinis reviewed (approved):
Did a small test and things look good 👍 
![image](https://github.com/user-attachments/assets/394855db-4b20-4e1e-8849-a1eeb7d37222)
![image](https://github.com/user-attachments/assets/2e7db1c8-a9c5-4859-80f1-2eea1e25949e)

tigrato commented on integrations/access/datadog/types.go:
```diff
@@ -0,0 +1,102 @@
+
+// Datadog API types
+
+type Metadata struct {
```
can you add documentation and describe where you got the information for these structures?

bernardjkim committed:
Document Datadog API types

bernardjkim commented:
Hey @tigrato @hugoShaka, are there any other major concerns with this PR? I will be make a couple follow up PRs, so we can always catch some minor issues then.

hugoShaka commented:
Yeah, the Enterprise tests are not passing and I'm trying to figure out why. (They are not run in the CI because the plugin is in Teleport OSS)

hugoShaka committed:
Only post resolution message when the AR is resolved

bernardjkim committed:
Fix lint

bernardjkim committed:
Unused function
